source/opt/log.h

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#ifndef SPIRV_TOOLS_LOG_H_
+#define SPIRV_TOOLS_LOG_H_
+
+// Logging levels for SPIRV-Tools in increasing verbosity order:
+// * SPIRV_LOG_LEVEL_FATAL:  Fatal error. Unrecoverable. Program will die
+//                           immediately. ALways enabled.
+// * SPIRV_LOG_LEVEL_ASSERT: Assertion failure. Something is wrong internally.
+//                           Program will die immediately. Disabled in
+//                           release build.
+// * SPIRV_LOG_LEVEL_DEBUG:  Debug information. Used for debugging purpose.
+//                           Disabled in release build.
+#define SPIRV_LOG_LEVEL_FATAL 0
+#define SPIRV_LOG_LEVEL_ASSERT 1
+#define SPIRV_LOG_LEVEL_DEBUG 2
+
+// By default we use SPIRV_LOG_LEVEL_ASSERT.
+#ifndef SPIRV_LOG_LEVEL
+#define SPIRV_LOG_LEVEL SPIRV_LOG_LEVEL_ASSERT
+#endif
+
+// TODO(antiagainst): add SPIRV_FATAL.
+
+// Asserts the given condition is true. Otherwise, print a message to stderr and
```

Comment:
For example, I think it's ugly that the unit tests print messages to stderr.  That shows bad encapsulation.